dialogue-annotations supports MultipartRequestBody.

jkozlowski opened the pull request:
## Before this PR
No built-in support for multipart.

## After this PR
==COMMIT_MSG==
dialogue-annotations provides support for MultipartRequestBody.
==COMMIT_MSG==

## Possible downsides?

* Implementation just wraps apache-httpclient things. I do not feel like we should invest the effort to reimplement the serialization.

jkozlowski committed:
Start plumbing a MultipartRequestBody.

jkozlowski commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/HttpEntityBodyRequestBodyAdapter.java:
```diff
@@ -0,0 +1,64 @@
+ * Adapts {@link HttpEntity} to a {@link RequestBody}.
+ */
+// Salt needs MIXED; other places use multipartBuilder.setType(MultipartBody.FORM);
+abstract class HttpEntityBodyRequestBodyAdapter implements RequestBody {
```
I'm thinking we should do the same trick with FormUrlEncoded?

carterkozak replied:
Don't we already have a pure form-encoded body implementation in WC?

jkozlowski replied:
yeah, that is defo a possibility. But thought might as well use the same trick for both of these. No preference though.

jkozlowski commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/MultipartRequestBody.java:
```diff
@@ -0,0 +1,97 @@
+        return new Builder();
+    }
+
+    public static RequestBodyPartBuilder requestBodyPartBuilder(RequestBody unsafeRequestBody) {
```
Not convinced RequestBody is the right interface here; one client provides InputStreams (roughly), another Files. We should think if supporting Content-Length makes much sense

jkozlowski replied:
In apache land, the repeatability is inferred by ```#getContentLength() != -1```

jkozlowski replied:
Fixed

jkozlowski committed:
Fixup test.

jkozlowski committed:
Remove comment.

jkozlowski committed:
Fix build.

jkozlowski commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/HttpEntityBodyRequestBodyAdapter.java:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Adapts {@link HttpEntity} to a {@link RequestBody}.
+ */
+abstract class HttpEntityBodyRequestBodyAdapter implements RequestBody {
```
Thought's on separate subproject for these? I guess Json is already here...

carterkozak replied:
I don't have a strong opinion either way -- I suppose the question is whether we'd expect these to ever be used without the dialogue annotation processor? I'd hope not, so this is a reasonable enough place for them to live.

jkozlowski replied:
I was going to call it dialogue-annotation-processor-extras or something, just to lighten the dependency. But then I realized that people already depend on apache, so it's a moot point. 

>  I'd hope not, so this is a reasonable enough place for them to live.

No, we do not want anyone to use this outside of annotation processor.

jkozlowski committed:
Hmmm

jkozlowski commented on dialogue-annotations/build.gradle:
```diff
@@ -12,6 +12,12 @@ dependencies {
     implementation 'com.google.code.findbugs:jsr305'
     implementation 'com.palantir.safe-logging:preconditions'
     implementation 'com.palantir.safe-logging:safe-logging'
+    implementation 'com.google.guava:guava'
```
Dunno why this needs to be here? I didn't add any guava things only to tests...

carterkozak replied:
huh, I'm not sure.

carterkozak replied:
huh, I'm not sure.

jkozlowski committed:
Start handling form bodies.

jkozlowski committed:
Mostly working test, except for Content-Length.

carterkozak commented:
Thoughts on a separate PR to update RequestBody with something like this? I considered it while working on the feign shim, but it was one-off enough that it wasn't worthwhile, but I think we have enough reason at this point.

```java
iface RequestBody {
default OptionalLong contentLength() {
  return OptionalLong.empty();
}
```

jkozlowski committed:
Fix build.

jkozlowski committed:
Fix test and Add ContentLength to RequestBody.

jkozlowski committed:
Start trying to plumb another test case.

jkozlowski committed:
Support PartMap.

jkozlowski committed:
Some more tests.

jkozlowski committed:
Custom ContentBody interface.

jkozlowski commented on dialogue-annotations/src/test/java/com/palantir/dialogue/annotations/MultipartRequestBodyTest.java:
```diff
@@ -0,0 +1,333 @@
+        String contentType();
+    }
+
+    private RequestBody unknownLengthRequestBody(byte[] value, okhttp3.MediaType contentType) {
```
So all this ceremony is because apache does not set Content-Length header for each individual part. as discussed with @carterkozak it doesn't feel like a problem, so we'll have it on our ContentBody interface, in case it's supported in the future, but unlikely to be a big deal for anyone that has to use this.

jkozlowski committed:
Final tweaks.

jkozlowski commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/ContentBody.java:
```diff
@@ -0,0 +1,42 @@
+    void writeTo(OutputStream output) throws IOException;
+
+    /** A HTTP content type (e.g., "application/json") indicating the type of content. */
+    String contentType();
```
Ok, so another quirk of apache implementation of this, is that #isRepeatable is based on getContentLength() != 1; interestingly we have a an internal client that deals with InputStreams that it may arbitrarily stream to disk, to make sure it can get retries. But I think that might be ameanable to providing the contentLength of those values, to get repeatability.

carterkozak replied:
We don't currently provide repeatability for `BinaryRequestBody` binary requests, it's reasonable not to support it in this case when lengths aren't provided -- I would not want to record contents to disk, as disk performance can be finicky, and space may not be available.

jkozlowski replied:
Yeah, I agree it's reasonable. Internal client likely was written because it was dealing with finnicky upstreams, so the way it deals with that is write to disk first, then send to itself. Anyway, feels like we can support what they need, so shouldn't be a big deal.

jkozlowski committed:
Util for Path -> ContentBody.
Test for repeatable.

jkozlowski committed:
Fix errorprone

jkozlowski committed:
Add #build to builders.

jkozlowski committed:
Bring back contentLength.

carterkozak commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/ContentBody.java:
```diff
@@ -0,0 +1,63 @@
+        return new com.palantir.dialogue.annotations.ContentBody() {
+            @Override
+            public void writeTo(OutputStream output) throws IOException {
+                Files.copy(filePath, output);
```
`contentLength`?

carterkozak replied:
Probably worth a similar factory for InputStream (non-repeatable)

jkozlowski replied:
I thought about it, but it felt racey? But I think there's enough precent to suggest this is fine.

jkozlowski replied:
Added and contentLength is pulled out of the File.

carterkozak commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/ContentBody.java:
```diff
@@ -0,0 +1,63 @@
+import java.nio.file.Path;
+import java.util.OptionalLong;
+
+public interface ContentBody extends Closeable {
```
We might as well add a `repeatable` flag while we're here rather than relying on contentLength as a sentinel.

jkozlowski replied:
Done

carterkozak commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/MultipartRequestBody.java:
```diff
@@ -0,0 +1,153 @@
+            super(Preconditions.checkNotNull(
+                    ContentType.parse(requestBody.contentType()),
+                    "Invalid content type",
+                    UnsafeArg.of("contentType", requestBody.contentType())));
```
content-type should be safe, even if invalid.

jkozlowski replied:
Done

carterkozak commented:
generally lgtm, just a couple comments

jkozlowski commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/ContentBody.java:
```diff
@@ -0,0 +1,63 @@
+     * preferring to catch and log.
+     */
+    @Override
+    void close();
```
I didn't want to break ranks here for reasons of consistency, but this feels weird, to have the logging in a bunch of places, as opposed to 1. Yes, the log message would be more generic, but the stacktrace should already capture everything you need to know about the type of resource, so not sure what we're getting by pushing this onto impls.

jkozlowski commented on dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/MultipartRequestBody.java:
```diff
@@ -0,0 +1,153 @@
+
+        @Override
+        public void writeTo(OutputStream out) throws IOException {
+            try (ContentBody requestBody = unsafeRequestBody) {
```
Just realized this isn't going to work so well, if we want repeatability.

jkozlowski replied:
Fixed.

jkozlowski committed:
Add generated changelog entries

jkozlowski committed:
Cleanup:
1. Repeatability is preserved and independent of content length.
2. Stop using abstract HttpEntityBodyRequestBodyAdapter, as there were tweaks that didn't make sense.
3. Correctly close parts, in a way that does not close them on just #writeTo.
4. Add InputStreamContentBody.
5. Extract explicit PathContentBody, and make it report length correctly.
6. Align variable names to contentBody.
7. Add a whole bunch of Preconditions.checkNotNull to tighten the API.

carterkozak commented on dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/ApacheHttpClientBlockingChannel.java:
```diff
@@ -177,24 +176,7 @@ private static boolean requiresEmptyBody(Endpoint endpoint) {
-                        SafeArg.of(HttpHeaders.CONTENT_LENGTH, contentLengthValue),
-                        nfe);
-            }
-        }
```
I think the feign shim relies on this logic, we should support Content-Length in both places at least for a few releases

jkozlowski replied:
Ack, I shall split a separate PR as discussed.

jkozlowski committed:
Merge branch 'develop' into jakubk/multipart

jkozlowski committed:
Merge branch 'develop' into jakubk/multipart

svc-autorelease commented:
Released [1.121.0](https://github.com/palantir/dialogue/releases/tag/1.121.0)
